datadog/fwprovider/resource_datadog_app_builder_app.go

```diff
@@ -0,0 +1,720 @@
+	if err != nil {
+		response.Diagnostics.AddError("error parsing ID as UUID", err.Error())
+		return
+	}
+
+	// delete the app
+	_, httpResp, err := r.Api.DeleteApp(r.Auth, id)
+	if err != nil {
+		if httpResp != nil {
+			// error body may have useful info for the user
+			body, err := io.ReadAll(httpResp.Body)
+			if err != nil {
+				response.Diagnostics.AddError("error reading error response", err.Error())
+				return
+			}
+			response.Diagnostics.AddError("error deleting app", string(body))
+		} else {
+			response.Diagnostics.AddError("error deleting app", err.Error())
+		}
+		return
+	}
+}
+
+func (r *appBuilderAppResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	// If the plan is null (resource is being destroyed) or no state exists yet, return early
+	// as there's nothing to modify
+	if req.Plan.Raw.IsNull() || req.State.Raw.IsNull() {
+		return
+	}
+
+	var plan, config, state appBuilderAppResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	resp.Diagnostics.Append(req.Config.Get(ctx, &config)...)
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// If app_json isn't set in either plan or config, nothing to do
+	if plan.AppJson.IsNull() || config.AppJson.IsNull() {
+		return
+	}
+
+	// Unmarshal all JSONs to compare their structures
+	var planJSON, configJSON, stateJSON map[string]interface{}
+	if err := json.Unmarshal([]byte(plan.AppJson.ValueString()), &planJSON); err != nil {
+		resp.Diagnostics.AddError("Error unmarshaling plan JSON", err.Error())
+		return
+	}
+	if err := json.Unmarshal([]byte(config.AppJson.ValueString()), &configJSON); err != nil {
+		resp.Diagnostics.AddError("Error unmarshaling config JSON", err.Error())
+		return
+	}
+	if err := json.Unmarshal([]byte(state.AppJson.ValueString()), &stateJSON); err != nil {
+		resp.Diagnostics.AddError("Error unmarshaling state JSON", err.Error())
+		return
+	}
+
+	// Helper function to recursively remove empty arrays, maps, and null values
+	// This ensures we don't treat empty/null values as meaningful differences
+	var removeEmptyValues func(map[string]interface{})
+	removeEmptyValues = func(m map[string]interface{}) {
+		for k, v := range m {
+			switch val := v.(type) {
+			case []interface{}:
+				if len(val) == 0 {
+					delete(m, k)
+				} else {
+					for _, item := range val {
+						if itemMap, ok := item.(map[string]interface{}); ok {
+							removeEmptyValues(itemMap)
+						}
+					}
+				}
+			case map[string]interface{}:
+				removeEmptyValues(val)
+				if len(val) == 0 {
+					delete(m, k)
+				}
+			case nil:
+				delete(m, k)
+			}
+		}
+	}
+
+	// Create copies of the JSON structures to modify without affecting originals
+	planCopy := make(map[string]interface{})
+	configCopy := make(map[string]interface{})
+	stateCopy := make(map[string]interface{})
+	maps.Copy(planCopy, planJSON)
+	maps.Copy(configCopy, configJSON)
```

Comment:
i don't think we should be modifying the config since that's basically the user input and messing with it could bypass TF's consistency rules. what's the reason it's needed here?

Reply:
ah good catch! this section of code is just copying for sake of comparison and not modifying the config itself. i think the config copies are leftover from previous iterations actually, since we end up only comparing the plan and state anyway. i'll remove the config stuff